Improve ABI types in wasmv1

bilboquet opened the pull request.

aoudiamoncef commented on proto/abis/massa/abi/v1/abi.proto:
```diff
@@ -82,22 +74,257 @@ message GenerateEventRequest {
   bool exists = 1;
 }
+
+// Call RespResult becaule Result will conflict with the Result type in Rust
```
Because

bilboquet replied:
x) 
(probably copilot fault ;)

aoudiamoncef reviewed (approved):
LGTM good job 👏 

Just add missing comments on messages and field and It's ready to ship

damip commented:
```
* use https://protobuf.dev/reference/csharp/api-docs/class/google/protobuf/well-known-types/empty instead of our own Empty

* use bytes in GenerateEvent

* SubNativeAmountsRequest
	use "left" and "right" and explain that we compute "left - right, fail if right > left" instead of "minuend/subtrahend"
 

* NativeAmountToBytesRequest
	my bad; technically the SC can already do byte conversion !

* remove LogRequest: we are deprecating log printing in favor of Events
```

bilboquet commented:
> LGTM good job clap
> 
> Just add missing comments on messages and field and It's ready to ship

Yes, it's still draft, because up to now I was still concerned by the structure of the .proto. (it was hardly extensible and not that easy to manipulate in Rust)
I think it's now ok, hence the PR :)
And so I can:
[] add comments
[] verify int vs fixed

bilboquet commented on proto/abis/massa/abi/v1/abi.proto:
```diff
@@ -82,22 +74,257 @@ message GenerateEventRequest {
 
 // TransferCoins
```
Those first requests are already fully implemented in the wasmv1. (we can write sc with them).
As testing is hard, I will first use them to tests amount, address… then I'll adjust them.
But yes most of them must be adapted so I keep this comment open ^^.

aoudiamoncef commented:
> 

IMHO, using well known type `Empty` could be a breaking change in the future. 

In the first version private API, I tried to use a xxxRequest xxxResponse even it's an empty. If we would like to change input/output of the method = no breaking change

bilboquet commented on proto/abis/massa/abi/v1/abi.proto:
```diff
@@ -13,65 +19,51 @@ option php_namespace = "Com\\Massa\\Abi\\V1";
-  fixed64 amount = 1;
+// Error message used in abi Response
+message Error {
+  string message = 1;
```
done

bilboquet commented:
> > LGTM good job clap
> > Just add missing comments on messages and field and It's ready to ship
> 
> Yes, it's still draft, because up to now I was still concerned by the structure of the .proto. (it was hardly extensible and not that easy to manipulate in Rust) I think it's now ok, hence the PR :) And so I can: 

[X] add comments 
[X] verify int vs fixed

bilboquet commented on proto/abis/massa/abi/v1/abi.proto:
```diff
@@ -13,65 +18,54 @@ option php_namespace = "Com\\Massa\\Abi\\V1";
+  bytes data = 1;
 }
 
 // GenerateEventRequest
```
use bytes in GenerateEvent but have the default SDK take strings and convert them to UTF8 when calling the ABI

bilboquet replied:
@damip 
just for information, the current wip implementation of this ABI in the AS SDK already does the conversion from UTF16 (user side) <--> UTF8 (abi side)
basically it's just adding a `to_vec`

damip replied:
Basically in the SC SDK we would take a utf16 string as argument and convert it to utf8 then to bytes before sending it to the abi

bilboquet replied:
That's what I was trying to say :)

bilboquet commented:
> * use https://protobuf.dev/reference/csharp/api-docs/class/google/protobuf/well-known-types/empty instead of our own Empty 

Own type removed, will use the one from google when needed (if the discussion in the other comment is correct)
[X] done

> * use bytes in GenerateEvent but have the default SDK take strings and convert them to UTF8 when calling the ABI 

added a dedicated comment to track this point (so consider this one closed)
[X] done

> * SubNativeAmountsRequest use "left" and "right" and explain that we compute "left - right, fail if right 
> left" instead of "minuend/subtrahend" 
[X] done

> * NativeAmountToBytesRequest technically the SC can already do to/from byte conversion ! I would vote for removing those to/from byte conversions and just use the builtin SC-based deserializer together with the Check * ABIs to verify the integrity of the data 

I took those 2 from the original discussion without much thinking... I fully agree that if the SC can do it by itself we can remove them. 
[X] done

> * remove LogRequest: we are deprecating log printing in favor of Events`

moved in a separate branch as this is very useful when writing the sdk, but for sure it should not be in the final ABI

bilboquet commented on proto/abis/massa/abi/v1/abi.proto:
```diff
@@ -13,65 +18,54 @@ option php_namespace = "Com\\Massa\\Abi\\V1";
-message Empty {}
-
-// CreateSC
 message CreateSCRequest {
```
Transform to the Response<Result,Error> form (later)

bilboquet commented on proto/abis/massa/abi/v1/abi.proto:
```diff
@@ -13,65 +18,54 @@ option php_namespace = "Com\\Massa\\Abi\\V1";
 }
 
 // CallSC
 message CallRequest {
```
DON'T Transform to the Response<Result,Error> form

bilboquet commented on proto/abis/massa/abi/v1/abi.proto:
```diff
@@ -13,65 +18,54 @@ option php_namespace = "Com\\Massa\\Abi\\V1";
 }
 
 // LocalCall
 message LocalCallRequest {
```
DON'T Transform to the Response<Result,Error> form

bilboquet commented on proto/abis/massa/abi/v1/abi.proto:
```diff
@@ -82,22 +76,319 @@ message GenerateEventRequest {
 
 // TransferCoins
 message TransferCoinsRequest {
```
Transform to the Response<Result,Error> form (later)

bilboquet commented on proto/abis/massa/abi/v1/abi.proto:
```diff
@@ -392,3 +399,107 @@ message DivRemNativeAmountResult {
 }
+
+// Comparison result
+message ComparisonResult {
```
@aoudiamoncef 
Do you think I should put this one in commons?
I feel like it should be

damip replied:
Yes I think so too, a bit like https://doc.rust-lang.org/std/cmp/enum.Ordering.html

bilboquet replied:
let's agree on the path before I move it
I propose 
`proto/commons/massa/shared/v1/comparison.proto`

it follow the scheme of others
`proto/abis/massa/abi/v1/abi.proto`
`proto/abis/massa/abi/v1/abi.proto`
`proto/commons/massa/model/v1/address.proto`...

bilboquet commented on proto/abis/massa/abi/v1/abi.proto:
```diff
@@ -82,22 +77,519 @@ message GenerateEventRequest {
+// keccak256 hash result
+message Keccak256Result {
+  // Hash of data
+  bytes hash = 1;
```
@damip 
What about returning a NativeHash?

bilboquet commented on proto/abis/massa/abi/v1/abi.proto:
```diff
@@ -13,91 +19,638 @@ option php_namespace = "Com\\Massa\\Abi\\V1";
+}
+
+// EVM signature verification request
+message VerifyEvmSigRequest {
```
For every Verify*SigRequest
put Sig as the first parameter

bilboquet commented on proto/abis/massa/abi/v1/abi.proto:
```diff
@@ -13,65 +19,51 @@ option php_namespace = "Com\\Massa\\Abi\\V1";
 }
 
-// CreateSCResponse
 message CreateSCResponse {
```
done
